### PR TITLE
Fix repair chain for tx indexing

### PR DESCRIPTION
**Pull request by aaronbuchwald**

### CODE REVIEW

**Comment by patrick-ogrady** [comment]
I think we should do this synchronously. It'll reduce lock contention (complete faster) and ensure we don't enter some weird state where we are repairing during active syncing.

File: core/blockchain.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ const (
 	maxTimeFutureBlocks  = 30
 	badBlockLimit        = 10
 	TriesInMemory        = 128
-	repairBlockBatchSize = 10
+	repairBlockBatchSize = 100
```
Comment (patrick-ogrady) [inline core/blockchain.go:94]: This is too big of a batch. I got this error locally:
```
DEBUG[02-05|19:28:42.110] repairing block                          hash=0xeac094d6e67457487fbf4d9d848d534ddef2feb59b6a2d1cdaee81a471db163e height=677
DEBUG[02-05|19:28:42.110] repairing block                          hash=0x58b1a904da9f39bfda81100aed616c1341c065a32a1400326750aecfcaf8f9e5 height=676
DEBUG[02-05|19:28:42.110] writing repair batch                     totalUpdates=48600 size=100
ERROR[02-05|19:28:42.180] failed to repair canonical chain         error="failed to write batch with size 0 at current block height 676: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (4195471 vs. 4194304)" timeElapsed=5m16.080370697s
```

**Comment by patrick-ogrady** [comment]
Got this error as well:
```
DEBUG[02-05|19:28:42.110] repairing block                          hash=0xeac094d6e67457487fbf4d9d848d534ddef2feb59b6a2d1cdaee81a471db163e height=677
DEBUG[02-05|19:28:42.110] repairing block                          hash=0x58b1a904da9f39bfda81100aed616c1341c065a32a1400326750aecfcaf8f9e5 height=676
DEBUG[02-05|19:28:42.110] writing repair batch                     totalUpdates=48600 size=100
ERROR[02-05|19:28:42.180] failed to repair canonical chain         error="failed to write batch with size 0 at current block height 676: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (4195471 vs. 4194304)" timeElapsed=5m16.080370697s
```

**Comment by patrick-ogrady** [comment]
Got this error when I moved batch size down to `20`:
```
DEBUG[02-05|19:43:14.561] repairing block                          hash=0x522c02c51e8887c6f22a0f62fc6ff2bc26ad13ee6269a0d32f76c53d9aed1221 height=715
DEBUG[02-05|19:43:14.562] repairing block                          hash=0xb098b53e4b876841e9293dab68f949eac0985a46ad620c68e38814cb47d2937f height=714
DEBUG[02-05|19:43:14.562] writing repair batch                     totalUpdates=48600 size=20
ERROR[02-05|19:43:14.629] failed to repair canonical chain         error="failed to write batch with size 0 at current block height 714: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (4195676 vs. 4194304)" timeElapsed=11m50.189963823s
```

**Comment by patrick-ogrady** [comment]
Current size here will always be 0: https://github.com/ava-labs/coreth/blob/52ac0d5d24f2c4210925e9409e53275696c07feb/core/blockchain.go#L946

File: plugin/evm/vm.go [hunk 2]
```diff
@@ -445,19 +445,21 @@ func (vm *VM) repairCanonicalChain() error {
...
+	if err := vm.chain.WriteCanonicalFromCurrentBlock(); err != nil {
+		return fmt.Errorf("failed to repair canonical chain after %v due to: %w", time.Since(start), err)
+	}
+	log.Info("finished repairing canonical chain", "timeElapsed", time.Since(start))
+	if err := vm.db.Put(repairedKey, []byte("finished")); err != nil {
+		return fmt.Errorf("failed to mark flag for canonical chain repaired due to: %w", err)
+	}
+	// TODO at this point, we should be in a valid state, so it may be alright to make this
+	// async
+	if err := vm.chain.ValidateCanonicalChain(); err != nil {
```
Comment (patrick-ogrady) [inline plugin/evm/vm.go:460]: Once we `WriteCanonicalFromCurrentBlock`, I think it is safe to assume we are in a good place?

Comment (aaronbuchwald) [reply plugin/evm/vm.go:460]: Yup, gonna convert this PR to a draft until I finalize this. This was just to check that was the case.

**Review by StephenButtolph — APPROVED** [review]

**Comment by patrick-ogrady** [comment]
Are we guaranteed the current block we are walking back from will be canonical?

File: core/blockchain.go [hunk 3]
```diff
@@ -895,57 +898,86 @@ func (bc *BlockChain) ValidateCanonicalChain() error {
 			}
 		}
 
-		current = bc.GetBlockByHash(current.ParentHash())
+		blkReceipts := bc.GetReceiptsByHash(current.Hash())
+		if blkReceipts.Len() != len(txs) {
```
Comment (patrick-ogrady) [inline core/blockchain.go:902]: We should confirm that:
* `len(receipts) == len(txs)`
* for each `receipt`, there is a matching block tx with the same hash
* each `receipt` has a block hash equal to the current block and and a block index equal to the current block

These could look something like: https://gist.github.com/patrick-ogrady/34e6458e698d5db68b8567abe7e96521#file-integrity_check-go-L98-L108

**Comment by patrick-ogrady** [comment]
This took ~27s when already synced to tip:
```
DEBUG[02-07|14:01:07.116] writing repair batch                     totalUpdates=54436 size=6
DEBUG[02-07|14:01:07.116] finished repairs                         totalUpdates=54436
INFO [02-07|14:01:07.116] finished repairing canonical chain       timeElapsed=27.14837944s
```

**Pull request merged**